From 9ac3ea4f6421ff331fedf5c181117b808714fe3c Mon Sep 17 00:00:00 2001
From: Bharat Bhushan <bbhushan2@marvell.com>
Date: Tue, 29 Oct 2019 12:04:11 +0530
Subject: [PATCH 392/767] perf/smmuv3: Fix programming event type with global

commit 439fc228868bd8f8b8f1cda07f343bf3579af58b from
git@git.assembla.com:cavium/WindRiver.linux.git
 filtering

Currently SMMU_PMCG_EVTYPERn.EVENT (event type) for n > 0
is not programmed on implementation supporting global filtering.
It uses default initialized values and reports counter of
default event-type rather than requested event-type.

Also SMMU_PMCG_EVTYPERn.SPAN and SMMU_PMCG_SMR(n) is RES0
for n > 0, so program these as zero for n > 0.

Change-Id: I2e9aad459b2b0a00977a0ab9080a2a83596c302f
Signed-off-by: Bharat Bhushan <bbhushan2@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/17961
Reviewed-by: Sunil Kovvuri Goutham <Sunil.Goutham@cavium.com>
Tested-by: Sunil Kovvuri Goutham <Sunil.Goutham@cavium.com>
Signed-off-by: Kevin Hao <kexin.hao@windriver.com>
---
 drivers/perf/arm_smmuv3_pmu.c | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/drivers/perf/arm_smmuv3_pmu.c b/drivers/perf/arm_smmuv3_pmu.c
index 5f8879f6cfd3..94f1e83bec2e 100644
--- a/drivers/perf/arm_smmuv3_pmu.c
+++ b/drivers/perf/arm_smmuv3_pmu.c
@@ -281,8 +281,13 @@ static int smmu_pmu_apply_event_filter(struct smmu_pmu *smmu_pmu,
 
 	/* Requested settings same as current global settings*/
 	if (span == smmu_pmu->global_filter_span &&
-	    sid == smmu_pmu->global_filter_sid)
+	    sid == smmu_pmu->global_filter_sid) {
+		if (idx == 0)
+			smmu_pmu_set_event_filter(event, idx, span, sid);
+		else
+			smmu_pmu_set_event_filter(event, idx, 0, 0);
 		return 0;
+	}
 
 	if (!bitmap_empty(smmu_pmu->used_counters, num_ctrs))
 		return -EAGAIN;
-- 
2.31.1

